File: src/leetcode/sept/day27.java
package leetcode.sept;

public class day27 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (p == null || q == null) return null;
        return dfs(root, p.val, q.val);
    }

    TreeNode dfs(TreeNode node, int p, int q) {

        if (node == null) return null;
        if (node.val == p || node.val == q) {
            return node;
        }
        if (node.val > p && node.val > q) {
            return dfs(node.left, p, q);
        } else if (node.val < p && node.val < q) {
            return dfs(node.right, p, q);
        } else return node;

    }
}

File: year2021/m4/day27.go
package m4

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rangeSumBST(root *TreeNode, low int, high int) int {

	if root == nil {
		return 0
	}
	res := 0
	if root.Val >= low && root.Val <= high {
		res = root.Val
	}
	if root.Val > low {
		res += rangeSumBST(root.Left, low, high)
	}
	if root.Val < high {
		res += rangeSumBST(root.Right, low, high)
	}
	return res
}
